### PR TITLE
Bfix/reorder imports

### PR DESCRIPTION
**Pull request by KarandikarMihir**
This is for #1229. Work's still in process.

### CODE REVIEW

**Comment by KarandikarMihir** [comment]
@gihrig Assuming that I'll work around these lines-
```
import 'file?name=[name].[ext]!./favicon.ico';
import '!file?name=[name].[ext]!./manifest.json';
import 'file?name=[name].[ext]!./.htaccess';
```

There's one problem that I can't wrap my head around why `import/first` rule doesn't accept imports like `components/A` or `containers/App`. I think this is a known issue (I'll refer to it once I find it).

**Comment by KarandikarMihir** [comment]
If you set import/first to `2` in `package.json`, you'd still find around 28 errors. All of them are related to [this (comment)](https://github.com/mxstbr/react-boilerplate/pull/1275#issuecomment-263116798)

**Comment by gihrig** [comment]
@KarandikarMihir I cloned your repo then added #1261 and I have no errors! 😄 

Check this out for yourself, but I think with #1261 merged you're PR is good to go. 

Let us know If this works out for you. If so, I'll try to squeeze out some time to test more thoroughly.

**Comment by KarandikarMihir** [comment]
@gihrig Awesome! Thanks for checking. Did you try setting `import/first` to `2`? I merged #1261 with this and it's still giving me errors. :confused:

**Comment by Dattaya** [comment]
So now we can get rid of this rule `"import/first": 0` altogether, because by default it is enabled in airbnb eslint config, right?

**Comment by gihrig** [comment]
@Dattaya Yes, I put it in there work around the errors you have fixed, it should be removed when you are done ✨

**Comment by gihrig** [comment]
@KarandikarMihir 
> The latest commit has zero errors with import/first set to 0.

How about with `import/first` removed?

**Comment by KarandikarMihir** [comment]
@gihrig No joy. It's giving me errors after removing `import/first`.

**Comment by gihrig** [comment]
@KarandikarMihir Do these new errors "make sense" in term of the order of imports you have been working on? In other words are they just more of the same or some weird new breed? 😉

**Comment by KarandikarMihir** [comment]
@gihrig They're the same ones! Nothing new, fortunately! 😌

Did you get a chance to look at `templates/app.js`? It's out of sync with `app/app.js`. The reason I'm insisting on this is because these files differ in the way they import some things. Are we even using `templates/app.js`?

**Comment by gihrig** [comment]
@KarandikarMihir Sorry I didn't respond to that question earlier. The files in `internals/templates` are used to create the app skeleton when the project is [`clean`](https://github.com/mxstbr/react-boilerplate/blob/master/docs/general/commands.md#cleaning)ed (note that they are also deleted when `clean` is run). So they will definitely _not_ match those of the example project and yes they too should be corrected as needed, specifically they need to be corrected so that the files they produce are error free.

**Comment by KarandikarMihir** [comment]
@gihrig That's okay! :smile: I know it's for generating skeleton app. What I meant to say is, those files are not identical.

See these lines in [template/app.js](https://github.com/mxstbr/react-boilerplate/blob/master/internals/templates/app.js#L103) and [app/app.js](https://github.com/mxstbr/react-boilerplate/blob/master/app/app.js#L120). They're not the same. Do we want them to be same?
I just happened to see this difference while reordering imports. Thought I might point it out.

**Comment by mxstbr** [comment]
That's correct @KarandikarMihir!

**Comment by KarandikarMihir** [comment]
@mxstbr Okay. So should I make changes to those files in this PR? Or is it a separate issue altogether?

**Comment by gihrig** [comment]
> So should I make changes to those files in this PR? Or is it a separate issue altogether?

As long as those changes are of a similar nature, just reorganizing code or minor refactorings that are not likely to affect execution, I would say go ahead and make them here. 

If they get too involved, open a new PR.

**Comment by KarandikarMihir** [comment]
I think it's done. Just need someone to thoroughly review the changes :)

**Comment by gihrig** [comment]
> I think it's done

Wow! that's fantastic ✨  I make time in the next 8 hrs or so to give it that thorough review ;-)

**Comment by KarandikarMihir** [comment]
@gihrig Yes that's the one! I'm pushing a new commit now.

**Comment by KarandikarMihir** [comment]
@gihrig Yes that's the one... I've removed `import/first` in the last commit, with `0` errors :smile: Thanks for all your help!!

**Comment by Dattaya** [comment]
> All of them are related to absolute path errors. How can containers/App be called as absolute path?

I asked your question in `eslint-plugin-import` repository: https://github.com/benmosher/eslint-plugin-import/issues/684, hopefully someone will answers it soon.

**Comment by gihrig** [comment]
Just to be clear, all the lint errors I see in the latest commit (93c5a9f23537b0e9298fb4af37b2cd7e80a16726) are `import/first` errors that can be resolved by re-organizing imports.

It's technically an error that `eslint-plugin-import` sees webpack aliases and _absolute path_ imports, but does this actually matter here?

For example, does anyone see a problem with this import order (it runs fine in limited testing).
```javascript
/**
 * app.js
 *
 * This is the entry file for the application, only setup and boilerplate
 * code.
 */

// Needed for redux-saga es6 generator support
import 'babel-polyfill';

// Import all the third party stuff
import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { applyRouterMiddleware, Router, browserHistory } from 'react-router';
import { syncHistoryWithStore } from 'react-router-redux';
import FontFaceObserver from 'fontfaceobserver';
import { useScroll } from 'react-router-scroll';
import 'sanitize.css/sanitize.css';

// Import selector for `syncHistoryWithStore`
import { selectLocationState } from 'containers/App/selectors';

// Load the favicon, the manifest.json file and the .htaccess file
/* eslint-disable import/no-webpack-loader-syntax */
import '!file?name=[name].[ext]!./favicon.ico';
import '!file?name=[name].[ext]!./manifest.json';
import 'file?name=[name].[ext]!./.htaccess'; // eslint-disable-line import/extensions
/* eslint-enable import/no-webpack-loader-syntax */

// Import Language Provider
import LanguageProvider from 'containers/LanguageProvider';

// Import root app
import App from 'containers/App';

import configureStore from './store';

// Import i18n messages
import { translationMessages } from './i18n';

// Import CSS reset and Global Styles
import './global-styles';

// Import routes
import createRoutes from './routes';
```

**Comment by KarandikarMihir** [comment]
@Dattaya Thanks for reporting the issue. Hopefully we'll have it fixed! @gihrig Yes, all the errors are only `import/first` related errors. So are you saying we should have `import/first: 0` in `package.json`?

**Comment by gihrig** [comment]
No, I'm saying remove the `import/first` rule from `package.json` and organize the imports so no errors are flagged.

**Comment by KarandikarMihir** [comment]
@gihrig Alright! I'll make some time for this.

**Comment by KarandikarMihir** [comment]
@gihrig Done and dusted!

**Comment by gihrig** [comment]
Fantastic 🎉 

99.99% there!

 I have tested _thoroughly_ and found only two issues in this monumental work.

Everything is good, until I generate a container. There are some issues in the generated code.

image

To reproduce:

```
npm run generate component TestComp /w defaults
npm run generate container TestPage /w defaults
npm run generate route TestPage - /test
use TestComp on TestPage
(patch generated test files just to make them pass)
npm test
```

And then there are merge conflicts... I'd like to call your attention to this [#1205 comment](https://github.com/mxstbr/react-boilerplate/pull/1205#issuecomment-265881061). Since this has been merged, `dev` is currently in a broken state, so beware of that.

I really appreciate your hanging in there with this, I imagine it turned out to be a bigger task than you expected from the start 😉

**Comment by KarandikarMihir** [comment]
@gihrig Thanks for the heads-up! That was tedious :sweat_smile: `generate`ing containers and components won't have any `import/first` errors. It does throw one lint error. 
```
/home/xxxx/react-boilerplate/app/containers/TestPage/sagas.js
  5:3  error  Unnecessary return statement  no-useless-return
```
I believe it's out of scope of this PR.

>I really appreciate your hanging in there with this, I imagine it turned out to be a bigger task than you expected from the start :wink:

IKR :goberserk:

**Comment by gihrig** [comment]
> I believe it's out of scope of this PR.

Yes, that issue is addressed in #1231

I'm having to end for the "day", I'll get back to testing this in about 9 hrs. 

Looking forward to merging!

File: internals/templates/app.js [hunk 1]
```diff
@@ -88,6 +94,7 @@ if (!window.Intl) {
   }))
     .then(() => Promise.all([
       System.import('intl/locale-data/jsonp/en.js'),
+      System.import('intl/locale-data/jsonp/de.js'),
```
Comment (mxstbr) [inline internals/templates/app.js]: Very good catch!

Comment (KarandikarMihir) [inline internals/templates/app.js]: Removed!

**Comment by gihrig** [comment]
Fantastic job! 🎉 

I have generated all options and there are no lint errors ✨ 
(excluding unrelated #1231 errors)

You sir, have made a great contribution to code quality, not just in RBP, but from here on _everyone_ will receive the benefit of `import/first`. My code quality is eternally grateful :100:

Merging.

**Comment by mxstbr** [comment]
Yes, thanks so so much @KarandikarMihir!

**Comment by KarandikarMihir** [comment]
@gihrig @mxstbr Thank you for sailing through thick and thin! :raised_hands: :innocent:

**Pull request merged**